File: src/main/java/com/viartemev/requestmapper/annotations/spring/extraction/PsiReferenceExpressionExtractor.java
package com.viartemev.requestmapper.annotations.spring.extraction;

import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiField;
import com.intellij.psi.PsiLiteralExpression;
import com.intellij.psi.PsiReferenceExpression;
import com.intellij.psi.impl.java.stubs.index.JavaStaticMemberNameIndex;
import com.intellij.psi.search.GlobalSearchScope;
import org.jetbrains.annotations.NotNull;

import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;

import static com.viartemev.requestmapper.utils.CommonUtils.unquote;
import static java.util.Objects.requireNonNull;
import static java.util.stream.Collectors.toList;

public class PsiReferenceExpressionExtractor implements PsiAnnotationValueExtractor<PsiReferenceExpression> {
    private final Project project;

    public PsiReferenceExpressionExtractor(Project project) {
        this.project = requireNonNull(project);
    }

    @Override
    public List<String> extract(@NotNull PsiReferenceExpression expression) {
        return Optional.ofNullable(expression.getReferenceName()).map(
                referenceName ->
                        JavaStaticMemberNameIndex.getInstance()
                             .get(referenceName,
                                  project,
                                  GlobalSearchScope.projectScope(project))
                             .stream()
                             .filter(PsiField.class::isInstance)
                             .map(PsiField.class::cast)
                             .flatMap(m -> Stream.of(m.getChildren()))
                             .filter(PsiLiteralExpression.class::isInstance)
                             .map(c -> unquote(c.getText()))
                             .collect(toList()))
                        .orElse(Collections.emptyList());

    }
}

File: src/main/kotlin/com/viartemev/requestmapper/annotations/spring/extraction/PsiReferenceExpressionExtractor.kt
package com.viartemev.requestmapper.annotations.spring.extraction

import com.intellij.psi.PsiBinaryExpression
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiLiteralExpression
import com.intellij.psi.PsiPolyadicExpression
import com.intellij.psi.PsiReferenceExpression
import com.viartemev.requestmapper.utils.unquote

class PsiReferenceExpressionExtractor : PsiAnnotationValueExtractor<PsiReferenceExpression> {

    override fun extract(value: PsiReferenceExpression): List<String> = listOf(extractPath(value))

    private fun extractPath(value: PsiReferenceExpression): String {
        return value.resolve()?.let {
            it
                    .children
                    .asSequence()
                    .filter { it is PsiBinaryExpression || it is PsiLiteralExpression || it is PsiPolyadicExpression }
                    .map { extractExpression(it) }
                    .toList()
                    //only one exists
                    .first()
        } ?: ""
    }

    private fun extractExpression(psiElement: PsiElement): String = when (psiElement) {
        is PsiLiteralExpression -> extractLiteralExpression(psiElement)
        is PsiBinaryExpression -> extractBinaryExpression(psiElement)
        is PsiPolyadicExpression -> extractPsiPolyadicExpression(psiElement)
        is PsiReferenceExpression -> extractPath(psiElement)
        else -> ""
    }

    private fun extractLiteralExpression(psiElement: PsiLiteralExpression) = psiElement.text.unquote()

    /** rOperand always presents in static final variables */
    private fun extractBinaryExpression(psiElement: PsiBinaryExpression) =
            extractExpression(psiElement.lOperand) + extractExpression(psiElement.rOperand!!)

    private fun extractPsiPolyadicExpression(psiElement: PsiPolyadicExpression) =
            psiElement
                    .operands
                    .joinToString(
                            separator = "",
                            transform = {
                                when (it) {
                                    is PsiLiteralExpression -> extractLiteralExpression(it)
                                    is PsiReferenceExpression -> extractPath(it)
                                    else -> ""
                                }
                            })
}